lib/DataLayoutAnalysis/Backend/DLAMakeModelTypes.cpp

```diff
@@ -455,76 +376,157 @@ static TypeMapT mapLLVMValuesToModelTypes(const LayoutTypeSystem &TS,
   return ValMap;
 }
 
-TypeMapT dla::makeModelTypes(const LayoutTypeSystem &TS,
-                             const LayoutTypePtrVect &Values,
-                             TupleTree<model::Binary> &Model) {
-  logEntry(TS, Model);
+static void makeTypeDefinitions(const dla::LayoutTypeSystem &TS,
+                                NodeToUpcastableTypeMap &NodesToTypes,
+                                TupleTree<model::Binary> &Model) {
 
-  const dla::VectEqClasses &EqClasses = TS.getEqClasses();
-  TypeVect Types;
-  Types.resize(EqClasses.getNumClasses());
-  PtrFieldsMap PointerFieldsToUpdate;
-
-  // Create nodes for anything that is not a pointer
-  llvm::SmallPtrSet<const LTSN *, 16> Visited;
-  for (const LTSN *Root : llvm::nodes(&TS)) {
-    revng_assert(Root != nullptr);
-    if (not isRoot(Root))
+  revng_log(Log, "makeTypeDefinitions");
+  LoggerIndent Indent{ Log };
+  for (const LTSN *Node : llvm::nodes(&TS)) {
+
+    if (not representsTypeDefinition(Node))
       continue;
 
-    for (const LTSN *N : llvm::post_order(ConstNonPointerFilterT(Root))) {
+    UpcastableType NewStructOrUnion = makeStructOrUnion(Model, Node);
+    revng_log(Log,
+              "Mapped Node with ID: "
+                << Node->ID << " to TypeDefinition with ID: "
+                << NewStructOrUnion->tryGetAsDefinition()->ID());
 
-      if (bool New = Visited.insert(N).second; not New)
-        continue;
+    NodesToTypes[Node] = NewStructOrUnion;
+  }
+}
 
-      auto T = createNodeType(Model,
-                              N,
-                              Types,
-                              EqClasses,
-                              PointerFieldsToUpdate);
+static void populateStructFields(const LayoutTypeSystemNode *StructNode,
+                                 model::StructDefinition *Struct,
+                                 const LayoutTypeSystem &TS,
+                                 NodeToUpcastableTypeMap &NodesToTypes,
+                                 TupleTree<model::Binary> &Model) {
+  LoggerIndent Indent{ Log };
+  for (const auto &[FieldNode, Edge] : StructNode->Successors) {
+    revng_log(Log, "Field ID: " << FieldNode->ID);
+    LoggerIndent FieldIndent{ Log };
+
+    revng_assert(TypeLinkTag::LK_Instance == Edge->getKind());
+
+    dla::OffsetExpression OE = Edge->getOffsetExpr();
+    UpcastableType
+      FieldType = makeFieldType<FieldKind::StructField>(FieldNode,
+                                                        OE,
+                                                        NodesToTypes,
+                                                        Model);
+    model::StructField Field{ OE.Offset, {}, {}, std::move(FieldType) };
```

Comment:
Using `{}` instead of `""` for `std::string` initialization is, I guess, technically more correct, but it always rubs me a wrong way a bit :D

Reply:
That being said, I wonder if we should just add this constructor, I feel like I have seen similar stuff elsewhere.

Reply:
Ah, wait! This should just be `Struct->addField(OE.Offset, std::move(FieldType))`.